File: solanaj/src/main/java/org/p2p/solanaj/kits/transaction/ConfirmedTransactionParsed.java
package org.p2p.solanaj.kits.transaction;

import com.google.gson.annotations.SerializedName;

import org.p2p.solanaj.model.types.RpcResultObject;

import java.util.List;
import java.util.Map;

public class ConfirmedTransactionParsed extends RpcResultObject {

    public static class AccountKeysParsed {
        @SerializedName("pubkey")
        private String pubkey;
        @SerializedName("signer")
        private boolean signer;
        @SerializedName("writable")
        private boolean writable;

        public String getPubkey() {
            return pubkey;
        }

    }

    public static class Parsed {
        @SerializedName("info")
        private Map<String, Object> info;
        @SerializedName("type")
        private String type;

        public Map<String, Object> getInfo() {
            return info;
        }

        public String getType() {
            return type;
        }

    }

    public static class InstructionParsed {
        @SerializedName("parsed")
        private Parsed parsed;
        @SerializedName("program")
        private String program;
        @SerializedName("programId")
        private String programId;
        @SerializedName("accounts")
        private List<String> accounts;
        @SerializedName("data")
        private String data;

        public Parsed getParsed() {
            return parsed;
        }

        public String gerProgram() {
            return program;
        }

        public String gerProgramId() {
            return programId;
        }

        public List<String> getAccounts() {
            return accounts;
        }

        public String getData() {
            return data;
        }

    }

    public static class MessageParsed {
        @SerializedName("accountKeys")
        private List<AccountKeysParsed> accountKeys;
        @SerializedName("instructions")
        private List<InstructionParsed> instructions;
        @SerializedName("recentBlockhash")
        private String recentBlockhash;

        public List<AccountKeysParsed> getAccountKeys() {
            return accountKeys;
        }

        public List<InstructionParsed> getInstructions() {
            return instructions;
        }

        public String gerRecentBlockhash() {
            return recentBlockhash;
        }

    }

    public static class TransactionParsed {
        @SerializedName("message")
        private MessageParsed message;
        @SerializedName("signatures")
        private List<String> signatures;

        public MessageParsed getMessage() {
            return message;
        }

        public List<String> getSignatures() {
            return signatures;
        }

    }

    public static class InnerInstruction {
        @SerializedName("index")
        private int index;

        @SerializedName("instructions")
        private List<InstructionParsed> instructions;

        public List<InstructionParsed> getInstructions() {
            return instructions;
        }

    }

    public static class PostTokenBalance {
        @SerializedName("accountIndex")
        private int accountIndex;
        @SerializedName("mint")
        private String mint;

        public int getAccountIndex() {
            return accountIndex;
        }

        public String getMint() {
            return mint;
        }

    }

    public static class Meta {
        @SerializedName("fee")
        private long fee;
        @SerializedName("innerInstructions")
        private List<InnerInstruction> innerInstructions;
        @SerializedName("postTokenBalances")
        private List<PostTokenBalance> postTokenBalances;

        public List<InnerInstruction> getInnerInstructions() {
            return innerInstructions;
        }

        public List<PostTokenBalance> getPostTokenBalances() {
            return postTokenBalances;
        }

        public long getFee() {
            return fee;
        }
    }

    @SerializedName("blockTime")
    private long blockTime;

    @SerializedName("slot")
    private int slot;

    @SerializedName("transaction")
    private TransactionParsed transaction;

    @SerializedName("meta")
    private Meta meta;

    public TransactionParsed getTransaction() {
        return transaction;
    }

    public int getSlot() {
        return slot;
    }

    public long getBlockTime() {
        return blockTime;
    }

    public Meta getMeta() {
        return meta;
    }

}

File: solanaj/src/main/java/org/p2p/solanaj/kits/transaction/ConfirmedTransactionParsed.kt
package org.p2p.solanaj.kits.transaction

import com.google.gson.annotations.SerializedName
import org.p2p.solanaj.model.types.RpcResultObject

data class ConfirmedTransactionParsed(
    @SerializedName("blockTime")
    val blockTime: Long,

    @SerializedName("slot")
    val slot: Int,

    @SerializedName("transaction")
    val transaction: TransactionParsed?,

    @SerializedName("meta")
    val meta: Meta,
) : RpcResultObject() {
    data class AccountKeysParsed(
        @SerializedName("pubkey")
        val pubkey: String?,

        @SerializedName("signer")
        private val signer: Boolean,

        @SerializedName("writable")
        private val writable: Boolean
    )

    data class Parsed(
        @SerializedName("info")
        val info: Map<String, Any>,

        @SerializedName("type")
        val type: String?
    )

    data class InstructionParsed(
        @SerializedName("parsed")
        val parsed: Parsed?,

        @SerializedName("program")
        val program: String?,

        @SerializedName("programId")
        val programId: String?,

        @SerializedName("accounts")
        val accounts: List<String>,

        @SerializedName("data")
        val data: String?
    )

    data class MessageParsed(
        @SerializedName("accountKeys")
        val accountKeys: List<AccountKeysParsed>,

        @SerializedName("instructions")
        val instructions: List<InstructionParsed>,

        @SerializedName("recentBlockhash")
        private val recentBlockhash: String
    )

    data class TransactionParsed(
        @SerializedName("message")
        val message: MessageParsed,

        @SerializedName("signatures")
        val signatures: List<String>
    )

    data class InnerInstruction(
        @SerializedName("index")
        private val index: Int,

        @SerializedName("instructions")
        val instructions: List<InstructionParsed>
    )

    data class TokenBalance(
        @SerializedName("accountIndex")
        val accountIndex: Int,

        @SerializedName("mint")
        val mint: String?
    )

    data class Meta(
        @SerializedName("fee")
        val fee: Long,

        @SerializedName("innerInstructions")
        val innerInstructions: List<InnerInstruction>,

        @SerializedName("postTokenBalances")
        val postTokenBalances: List<TokenBalance>,

        @SerializedName("preTokenBalances")
        val preTokenBalances: List<TokenBalance>
    )
}
